### ISSUE TITLE
Running Django 3 with asgi fails

### ISSUE THREAD

**Issue by blakerouse** [elastic/apm-agent-python#737]
**Describe the bug**:

Django 3 supports running with asgi. APM does not work when using asgi because an issue with the format of headers in the ASGI. In ASGI the headers are a list where APM current expect it to be a dictionary.

**To Reproduce**

1. Create a new Django 3 project.
2. Run uvicorn pointing at `newapp.asgi`
3. Get stacktrace:

```
[2020-02-25 19:16:15 +0000] [255] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/uvicorn/protocols/http/httptools_impl.py", line 385, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/usr/local/lib/python3.6/dist-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.6/dist-packages/myproject/asyncio.py", line 56, in __call__
    return await self.application(scope, receive, send)
  File "/usr/local/lib/python3.6/dist-packages/django/core/handlers/asgi.py", line 155, in __call__
    await sync_to_async(signals.request_started.send)(sender=self.__class__, scope=scope)
  File "/usr/local/lib/python3.6/dist-packages/asgiref/sync.py", line 244, in __call__
    return await asyncio.wait_for(future, timeout=None)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 339, in wait_for
    return (yield from fut)
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.6/dist-packages/asgiref/sync.py", line 277, in thread_handler
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/contextvars/__init__.py", line 38, in run
    return callable(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/django/dispatch/dispatcher.py", line 175, in send
    for receiver in self._live_receivers(sender)
  File "/usr/local/lib/python3.6/dist-packages/django/dispatch/dispatcher.py", line 175, in <listcomp>
    for receiver in self._live_receivers(sender)
  File "/usr/local/lib/python3.6/dist-packages/elasticapm/contrib/django/apps.py", line 144, in _request_started_handler
    trace_parent = TraceParent.from_headers(kwargs["scope"]["headers"])
  File "/usr/local/lib/python3.6/dist-packages/elasticapm/utils/disttracing.py", line 96, in from_headers
    tracestate = cls.merge_duplicate_headers(headers, tracestate_header_name)
  File "/usr/local/lib/python3.6/dist-packages/elasticapm/utils/disttracing.py", line 121, in merge_duplicate_headers
    return headers.get(key)
AttributeError: 'list' object has no attribute 'get'
127.0.0.1:52032 - "GET /healthz HTTP/1.0" 500
```

**Expected behavior**:

Spans are sent to APM server.

**Environment (please complete the following information)**
- OS: Ubuntu 18.04 (Docker)
- Python version: 3.6
- Framework and version: Django  3.0.1
- APM Server version: 7.6.0
- Agent version: 5.4.2

**Additional context**

Currently running under gunicorn in production but running locally with uvicorn gets the same result:

```
"""Settings for running gunicorn."""

bind = "0.0.0.0:8001"
workers = 4
worker_class = "uvicorn.workers.UvicornWorker"

accesslog = "-"
errorlog = "-"
loglevel = "info"

user = "www-data"
group = "www-data"

secure_scheme_headers = {"X-FORWARDED-PROTOCOL": "ssl", "X-FORWARDED-PROTO": "https"}
forwarded_allow_ips = "*"
```

**Comment by blakerouse** [comment]
I added a workaround in Django settings file:

```
from elasticapm.utils import disttracing

# Fix disttracing.TraceParent from elaticapm
def asgi_merge_duplicate_headers(headers, key):
    if isinstance(headers, list):
        return ",".join([
            item[1]
            for item in headers
            if item[0] == key
        ])
    return headers.get(key)

disttracing.TraceParent.merge_duplicate_headers = asgi_merge_duplicate_headers
```

**Comment by blakerouse** [comment]
I had to do more workarounds to get this to work. Django with ASGI has a few issues. The one I documented above along with contextvar handling. The usage of django signals for the start and finish of a request is an issue with ASGI.

django/asgiref run those signals in a ThreadPool so they lose context. I had to do the following as patches to get it to work with ASGI. But it does work!

```
"""APM Patching.

Patches Elastic APM to work with Django asgi.
"""

from functools import partial

from elasticapm.contrib.django import apps, middleware
from elasticapm.utils import disttracing


def asgi_merge_duplicate_headers(headers, key):
    """ASGI headers are a list of tuples."""
    if isinstance(headers, list):
        return ",".join([item[1] for item in headers if item[0] == key])
    return headers.get(key)


def register_handlers(client):
    """Registers the handlers.

    This is different from the version from Elastic APM, being that it doesn't
    use the request start/finish signals. When running with asgi those signals
    are ran on a threadpool from asyncio so the contextvars does not work
    correctly.
    """
    from django.core.signals import got_request_exception
    from elasticapm.contrib.django.handlers import exception_handler

    # Connect to Django's internal signal handlers
    got_request_exception.disconnect(dispatch_uid=apps.ERROR_DISPATCH_UID)
    got_request_exception.connect(
        partial(exception_handler, client),
        dispatch_uid=apps.ERROR_DISPATCH_UID,
        weak=False,
    )

    # If we can import celery, register ourselves as exception handler
    try:
        import celery  # pylint:disable=unused-import
        from elasticapm.contrib.celery import register_exception_tracking

        try:
            register_exception_tracking(client)
        except Exception as exc:  # pylint:disable=broad-except
            client.logger.exception("Failed installing django-celery hook: %s" % exc)
    except ImportError:
        client.logger.debug("Not instrumenting Celery, couldn't import")


# Patch merge_duplicate_headers to work with asgi.
disttracing.TraceParent.merge_duplicate_headers = asgi_merge_duplicate_headers

# Patch the register handlers.
apps.register_handlers = register_handlers


class ASGITracingMiddleware(middleware.TracingMiddleware):
    """Patch Elastic APM `Tracing Middleware` to work with Django 3 asgi.

    **NOTE**: Must be inserted as the first middleware manually.
    """

    def process_request(self, request):
        """Call request started."""
        apps._request_started_handler(self.client, None, request)

    def process_response(self, request, response):
        """Finalize the request."""
        response = super().process_response(request, response)
        if apps._should_start_transaction(self.client):
            self.client.end_transaction()
        return response
```

The reason this works is because the middleware are always executed on the same thread where as the signals will not be. This is specific to running Django with ASGI.

**Comment by basepi** [comment]
Thank you for your great detective work here @blakerouse! I'm going to spend some time tomorrow looking over your code changes and figure out the best way to include these fixes in the codebase.

**Comment by basepi** [comment]
@blakerouse Quick update here.

`process_request` and `process_response` are part of the [old-style Django middleware](https://django.cowhite.com/blog/django-middleware-the-new-version-of-middleware-and-how-to-use-it/), so our solution will definitely not use those (as they are not present starting in Django 2.0). But that's not a blocker for creating the transaction in the middleware, it's just a side note.

The bigger problem with starting the transaction in the middleware is that we are then unable to measure the time spend in the resolver and in any middleware that is before ours. The signals are early enough in the request that we're able to capture all of that additional data.

That said, I'd much rather lose contextual data than the actual transaction itself, due to separate thread contexts. I'm still evaluating potential solutions here, so stay tuned.

**Comment by basepi** [comment]
(Oh, the `process_request` and `process_response` do work in our code due to `django.utils.deprecation.MiddlewareMixin` -- you can ignore that note above)

**Comment by blakerouse** [comment]
Another possible way is to add a custom ASGI middleware, but the issue would be to set current transaction in the contextvar for the selected executed thread when asyncio runs the request through the middleware.

**Comment by basepi** [comment]
I actually have a hunch that this may be a python 3.6 issue. From what I can tell, for `contextvars`, this [has already been fixed](https://github.com/django/asgiref/issues/71). However, python 3.6 doesn't support `contextvars`. I'll be working on reproducing this tomorrow and testing that hypothesis, but if you'd like to get a head start you can try installing [the backported `contextvars`](https://pypi.org/project/contextvars/) and seeing if that changes anything. I'm also planning to test against python 3.7 since the 3.6 backport has some asterisks around async support.

**Comment by basepi** [comment]
I think my hypothesis is wrong. It doesn't appear to work with `contextvars` installed on 3.6.10, nor with 3.7.6.

Back to the drawing board.

**Comment by basepi** [comment]
We discovered (or, more precisely, @beniwohli discovered) that the `contextvars` fix that I linked to above used a copy of the `contextvars`, and thus our transaction `ContextVar` is being lost. I [submitted a PR](https://github.com/django/asgiref/pull/148) to asgiref to get this fixed, though we won't be able to benefit much from it due to supporting so many versions of Django.

With the fix above in place it works (though would still require `contextvars` to be installed on python 3.6). Of course, we still need an interim solution since that won't be available until it's released, and also won't be backported. I'm working on that now.

**Comment by basepi** [comment]
My [fix](https://github.com/django/asgiref/pull/148) to `asgiref` has been merged, and after a bunch more experimentation, I think we're going to rely on that fix rather than trying to hack our way around it. None of the solutions which I came up with were ideal, and as soon as the new version of `asgiref` is released it should be pretty trivial to upgrade.

If you want to test to make sure my fix works, you'll want to install `asgiref` from its master branch, and also install `contextvars` (since it's not included in python 3.6).
